Separated out functionalities from initPopulation

aahaselgrove opened the pull request:
`initPopulation` was a heavily overloaded parameter, with a variety of
types corresponding to a range of possible behaviours. This commit:
- Removes the option to pass a vector representing the search space.
This is specific and unclear, and behaviour can easily be easily
represented by one of the other options
- Restricts `initPopulation` to be a vector of individuals (where an
individual is a single member of the population)
- Adds a new parameter `creation` to represent a function to create an
individual.

Default behaviour has not been changed, but is now represented by a
creation function.

wildart commented on src/Evolutionary.jl:
```diff
@@ -17,7 +17,6 @@ using Random
            es, cmaes, ga
 
     const Strategy = Dict{Symbol,Any}
-    const Individual = Union{Vector, Matrix, Function, Nothing}
```
Why did you drop this? `Matrix` type can be used for fixed size population. Removing `Function` would be enough as you introduce new parameter for functional creation.

aahaselgrove replied:
I didn't find the 'Individual' type to be easy to understand - I don't think it represents an 'individual' at all, but rather the type of initialiser. As this PR breaks down initialisation into two parameters, one which must be a function, and one which must be a vector, I felt that the union type was no longer necessary.

wildart commented on src/es.jl:
```diff
@@ -11,13 +11,14 @@
-              mutation::Function = ((r,m)->r),
-              smutation::Function = (s->s),
-              termination::Function = (x->false),
+              creation::Function = (n -> rand(n)),
```
Now, the `creation` function could  generate vector with the type different from `initPolulation` if only `initPolulation` parameter passed.

aahaselgrove replied:
My intent here is that either a creation function or an initPopulation vector is passed, and if both are passed, the creation function is not called. Unfortunately Julia functions are not parameterised by argument or return types, so we cannot enforce creation (or for that matter, any other function used) to accept or return compatible types.

aahaselgrove commented:
The reason I created this PR is because I was primarily working towards my second PR https://github.com/wildart/Evolutionary.jl/pull/37 which addresses my issue https://github.com/wildart/Evolutionary.jl/issues/35, and I was having difficulty accommodating the existing initialisation system. Specifically, for populations with individuals which aren't Vector{T} (eg. object individuals or higher dimension individuals), Matrix{T} is not appropriate for initialising a group of individuals. Hence, I changed it to Vector{I}, which could be Vector{Vector{T}}, Vector{Object} or Vector{Array{T, N}} accordingly.

With regards to adding a creation function, I added this because it felt weird to me for initPopulation to be a vector or a function. Also, having separate parameters means that, in future, additional functionality could easily be implemented for passing partial initial populations: for an initPopulation of length less than N, the creation function could be used to generate the remaining individuals.

wildart commented:
> Specifically, for populations with individuals which aren't Vector{T} (eg. object individuals or higher dimension individuals), Matrix{T} is not appropriate for initialising a group of individuals. Hence, I changed it to Vector{I}, which could be Vector{Vector{T}}, Vector{Object} or Vector{Array{T, N}} accordingly.

The `Individual` type does not specify any type parameter on `Vector`, so essentially individual can be vector of anything: numbers, array, object, matrix, anything. `Matrix` type is used to introduce a predefined population.

I like a separate parameter for a creation function, so I think it would be better to set a following initialization procedure for population as follows:

    If no creation function specified
       if initPopulation is Matrix
          convert initial population Matrix to Vector{Vector} which is an internal population data structure
       else if element is initPopulation parameter is of primitive type
          generate Vector of random weights multiplied to initPopulation
       else
          error & advise to use creation function
    else // creation function provided
       if initPopulation is set
          warn that this parameter will be disregarded
       initialize population with elements generated from creation function

In the above initialization procedure, there is no need match types between initial population parameter and creation function output.

aahaselgrove commented:
> In the above initialization procedure, there is no need match types between initial population parameter and creation function output.

I don't quite follow. In this PR, there is no need to match types between initial population and creation function either. The creation function is never called if a value is provided for initial population.

> The `Individual` type does not specify any type parameter on `Vector`, so essentially individual can be vector of anything: numbers, array, object, matrix, anything. `Matrix` type is used to introduce a predefined population.

I'll add matrix back in here but remove it in later PRs because it isn't generic enough for populations of types other than Vector{T}

wildart commented:
Clearly, we got some misunderstanding. Let's get to the beginning. So, what do you think initial population parameter should be?

aahaselgrove commented:
For my project, I am working off aahaselgrove/Evolutionary.jl@0cd421d, in which the function signature for `ga` is as follows:

https://github.com/wildart/Evolutionary.jl/blob/0cd421d0c206ded341cc458edef01e6e15275ce0/src/ga.jl#L1-L31

Hence, default behaviour is for the initial population to be generated by the default creation function `(dims -> rand(eltype(T), dims))`.

wildart commented:
The creation function `(dims -> rand(eltype(T), dims))` will **only** work for primitive types. Either you need to specialize type parameter  as `where {T <: Real}` or come-up with a different initialization procedure.

Can you start putting some kind of general initialization procedure for the population from initial parameters for GA, and then try to reuse in in other methods?

wildart commented:
I like that in your ` ga(objfun::Function, individual::T)`, you pass individual as a positional parameter, that makes it easy to perform multiple dispatch on other form of individuals. Maybe you can rewrite your current PR in that way.

wildart commented:
Here is my take on initialization:
- Add `population` parameter that  will hold a working population
https://github.com/wildart/Evolutionary.jl/blob/d420050869bcc522c538628d87629ad1a3e1de2d/src/es.jl#L17
- Dispatch on various population initialization parameters 
https://github.com/wildart/Evolutionary.jl/blob/d420050869bcc522c538628d87629ad1a3e1de2d/src/es.jl#L109-L125

In that way, population initialization is moved out of the main algorithm, and the new interface would allow to provide special initialization functions for custom population creation procedure.

aahaselgrove commented:
> Here is my take on initialization:

I like the multiple dispatch approach for population initialization - it feels appropriate to the language. I still have problems with the options of individuals you provide for the reasons I outlined above.

> The creation function `(dims -> rand(eltype(T), dims))` will **only** work if **T** is primitive type. Either you need to specialize type parameter as `where {T <: Real}` or come-up with a different initialization procedure.

Your method of creation also only works only if T is primitive. I don't think it's feasible to be generic enough to cover all kinds of population, and it is reasonable to require the user to perform initialisation for more complex types. However, I could certainly see it being appropriate to throw an error if the user tries to run ga with a more complex individual without defining a creation function or passing an initial population.

aahaselgrove commented:
Also, can `lowerBounds` and `upperBounds` be removed? They don't do anything at the moment.

wildart commented:
> Also, can `lowerBounds` and `upperBounds` be removed? They don't do anything at the moment.

Really, I thought they are functional. Anyway, leave them for future development.

> I believe the existing method of creation also only works only if T is primitive.

Surely, the default `creation` parameter is not going to work custom types, but with an appropriate creation function, initialization will work.

> it is reasonable to require the user to perform initialization for more complex types

That's why the Matrix or Vector{Vector} population parameters were provided - to initialize whole population in advance.

aahaselgrove commented:
I think I've hit an effective compromise now. Initialisation can still be completed by passing a vector representing the search space or a matrix, but this is no longer part of the core algorithm, but more like helper functions for common initialisation patterns. The core algorithm is to first use any individuals passed in with the `initPopulation` vector, and create any remaining individuals with the creation function (if `populationSize > length(initPopulation)`).

wildart commented:
>  Initialisation can still be completed by passing a vector representing the search space or a matrix, but this is no longer part of the core algorithm

I think it's a good thing to move out initialization out of core algorithm. Do not stop in half way move all initializations into the separate function.

> The core algorithm is to first use any individuals passed in with the `initPopulation` vector, and create any remaining individuals with the creation function (if `populationSize > length(initPopulation)`).

Mixing two types of the population initialization isn't a very nice thing. It could create ambiguities down the line.

aahaselgrove commented:
> Mixing two types of the population initialization isn't a very nice thing. It could create ambiguities down the line.

Do you have any examples of how this would be an issue? MATLAB's `ga` function works in exactly this manner, allowing the user to specify as many or as few individuals as they want in the initial population and then creates additional as required with the creation function for a total of populationSize individuals.

wildart commented:
> Do you have any examples of how this would be an issue?

Well, you can have population values type-incompatible with the creation function produced individuals. I looked at MATLAB `ga` parameter description, you were right they allow such scenario. I guess it's alright to have such behavior.

Nevertheless, could you move any initialization code out of the `ga` main function. I plan to refactor code to have clear functional separations of various parts of the algorithms. That will allow to start work on concurrent implementation.

aahaselgrove commented:
> Well, you can have population values type-incompatible with the creation function produced individuals.

Population values could also be incompatible with `objfun`, `crossover` or `mutation`.

> Nevertheless, could you move any initialization code out of the `ga` main function.

I'll take a look at this.

wildart commented:
Implemented in new api, see [Docs/Dev/Population](https://wildart.github.io/Evolutionary.jl/dev/dev/#Population-1)
